src/native/containers/dn-simdhash-specialization.h

```diff
@@ -0,0 +1,524 @@
+#define DN_SIMDHASH_INSTANCE_DATA_T DN_SIMDHASH_T_PTR
+#else // DN_SIMDHASH_INSTANCE_DATA_T
+#define DN_SIMDHASH_GET_DATA(hash) dn_simdhash_instance_data(DN_SIMDHASH_INSTANCE_DATA_T, hash)
+#endif // DN_SIMDHASH_INSTANCE_DATA_T
+
+#ifndef DN_SIMDHASH_KEY_HASHER
+#error Expected DN_SIMDHASH_KEY_HASHER definition with signature: uint32_t (DN_SIMDHASH_INSTANCE_DATA_T data, KEY_T key)
+#endif
+
+#ifndef DN_SIMDHASH_KEY_EQUALS
+#error Expected DN_SIMDHASH_KEY_EQUALS definition with signature: int (DN_SIMDHASH_INSTANCE_DATA_T data, KEY_T lhs, KEY_T rhs) that returns 1 for match
+#endif
+
+#ifndef DN_SIMDHASH_ON_REPLACE
+#define DN_SIMDHASH_HAS_REPLACE_HANDLER 0
+#define DN_SIMDHASH_ON_REPLACE(data, key, old_value, new_value)
+#else // DN_SIMDHASH_ON_REPLACE
+#define DN_SIMDHASH_HAS_REPLACE_HANDLER 1
+#ifndef DN_SIMDHASH_ON_REMOVE
+#error Expected DN_SIMDHASH_ON_REMOVE(data, key, value) to be defined.
+#endif
+#endif // DN_SIMDHASH_ON_REPLACE
+
+#ifndef DN_SIMDHASH_ON_REMOVE
+#define DN_SIMDHASH_HAS_REMOVE_HANDLER 0
+#define DN_SIMDHASH_ON_REMOVE(data, key, value)
+#else // DN_SIMDHASH_ON_REMOVE
+#define DN_SIMDHASH_HAS_REMOVE_HANDLER 1
+#ifndef DN_SIMDHASH_ON_REPLACE
+#error Expected DN_SIMDHASH_ON_REPLACE(data, key, old_value, new_value) to be defined.
+#endif
+#endif // DN_SIMDHASH_ON_REMOVE
+
+#ifndef DN_SIMDHASH_BUCKET_CAPACITY
+// TODO: Find some way to automatically select an ideal bucket capacity based on key size.
+// Some sort of trick using _Generic?
+#define DN_SIMDHASH_BUCKET_CAPACITY DN_SIMDHASH_DEFAULT_BUCKET_CAPACITY
+#endif
+
+#include "dn-simdhash-specialization-declarations.h"
+
+static_assert(DN_SIMDHASH_BUCKET_CAPACITY <= DN_SIMDHASH_MAX_BUCKET_CAPACITY, "Maximum bucket capacity exceeded");
+static_assert(DN_SIMDHASH_BUCKET_CAPACITY > 1, "Bucket capacity too low");
+
+// We set bucket_size_bytes to sizeof() this struct so that we can let the compiler
+//  generate the most optimal code possible when we're manipulating pointers to it -
+//  that is, it can do mul-by-constant instead of mul-by-(hash->meta.etc)
+// We use memcpy to do an unaligned load when reading dn_simdhash_suffixes, but it's
+//  still ideal to align instances of bucket_t to match the vector width, so that
+//  loads are less likely to span two cache lines.
+#ifdef _MSC_VER
+typedef struct __declspec(align(DN_SIMDHASH_VECTOR_WIDTH)) bucket_t {
+#else
+typedef struct bucket_t {
+#endif
+	dn_simdhash_suffixes suffixes;
+	DN_SIMDHASH_KEY_T keys[DN_SIMDHASH_BUCKET_CAPACITY];
+}
+#if defined(__clang__) || defined(__GNUC__)
+__attribute__((__aligned__(DN_SIMDHASH_VECTOR_WIDTH))) bucket_t;
+#else
+bucket_t;
+#endif
+
+static_assert((sizeof (bucket_t) % DN_SIMDHASH_VECTOR_WIDTH) == 0, "Bucket size is not vector aligned");
+
+
+// While we've inlined these constants into the specialized code we're generating,
+//  the generic code in dn-simdhash.c needs them, so we put them in this meta header
+//  that lives inside every hash instance. (TODO: Store it by-reference?)
```

Comment:
> TODO: store it by-reference

Aren't you already storing it by reference?

> ```c
>     dn_simdhash_meta_t *meta;
> ```

Reply:
Thanks, the comment is outdated now